apps/backend/src/parse/parse.service.ts

```diff
@@ -254,76 +249,103 @@ export class ParseService {
-      const finalParseList = parseList.filter(
-        (filename) => !filename?.includes('LABOUR2')
+    this.appLogger.log(`Found ${finalParseList.length} files to parse...`);
+    return finalParseList;
+  }
+  /**
+   * If no records have been passed in from the event, we will check the S3 bucket for any unparsed files
+   * @returns
+   */
+  async checkS3ForFiles(): Promise<string[]> {
+    try {
+      return (
+        (await this.s3.listBucketContents(
+          `pcc-integration-data-files-${process.env.RUNTIME_ENV}`
+        )) ?? []
       );
-
-      this.appLogger.log(`Found ${finalParseList.length} files to parse...`);
-
-      // Parse & Save only files that have not been parsed before
-      for (const filename of finalParseList) {
-        this.appLogger.log(`Parsing ${filename}..`);
-        if (filename) {
-          await this.parseFileFromS3(filename);
-        }
-      }
     } catch (err) {
       this.appLogger.error(err);
+      throw new Error('Error checking S3 for files');
     }
   }
-
-  async parseFileFromS3(fileKey: string): Promise<FileUploadedEntity | void> {
+  /**
+   * Find the program rules and validates the file
+   * @param filename
+   * @param program
+   * @returns
+   */
+  async validateFileList(filename: string, program: Ministries) {
     try {
-      const bucket = `pcc-integration-data-files-${process.env.RUNTIME_ENV}`;
-
-      //file source is either bcm, or the txn data from a program
-      const fileSource = fileKey.split('/')[0];
-      const filename = fileKey.split('/')[1];
-      // program is the ministry - there will be txn fileSource data and bcm fileSource data for each ministry
-      const file = await this.s3.getObjectString({
-        Bucket: bucket,
-        Key: `${fileSource}/${filename}`,
-      });
+      const allFiles: FileUploadedEntity[] = await this.getAllFiles();
+      const allFilenames = new Set(allFiles.map((f) => f.sourceFileName));
+
+      if (allFilenames.has(filename)) {
+        throw new BadRequestException({
+          message: 'Invalid filename, this already exists',
+        });
+      }
 
       //Get all existing rules for each program
       const rules: FileIngestionRulesEntity[] =
         await this.notificationService.getAllRules();
 
+      // Throws an error if no rules exist for the specified program
+      const programRules: FileIngestionRulesEntity =
+        await this.notificationService.getRulesForProgram(program);
+
+      if (!rules) {
+        // TODO email? Do we include the message in the case of HTTP exception?
+        throw new HttpException(
+          `No rules established for program ${program}`,
+          HttpStatus.FORBIDDEN
+        );
+      }
+
       const currentRule: FileIngestionRulesEntity | undefined = rules.find(
         (rule) => filename.includes(rule.program)
       );
 
       if (currentRule === undefined) {
+        // TODO email? Do we include the message in this case?
```

Comment:
I say no. This would be more of an internal error than something going out to the suer.